src/hotspot/share/gc/parallel/psOldGen.cpp

```diff
@@ -163,14 +168,16 @@ bool PSOldGen::expand_for_allocate(size_t word_size) {
   assert(word_size > 0, "allocating zero words?");
   bool result = true;
   {
-    MutexLocker x(ExpandHeap_lock);
+    MutexLocker x(&_expand_lock);
     // Avoid "expand storms" by rechecking available space after obtaining
     // the lock, because another thread may have already made sufficient
     // space available.  If insufficient space available, that will remain
     // true until we expand, since we have the lock.  Other threads may take
     // the space we need before we can allocate it, regardless of whether we
     // expand.  That's okay, we'll just try expanding again.
-    if (object_space()->needs_expand(word_size)) {
+    bool needs_expand =
+      pointer_delta(object_space()->end(), object_space()->top()) < word_size;
+    if (needs_expand) {
```

Comment:
To me the old code reads better, but I guess it's a matter of taste. The predicate could be moved to PSOldGen to allow asserting that the lock is held.